virtwho/register.py

```diff
@@ -1,5 +1,487 @@
+        for i in range(3):
+            self.unregister(guest=guest)
+            ret, output = ssh.runcmd(cmd)
+            if ret == 0 and 'The system has been registered' in output:
+                logger.info(f'Succeeded to register {host}')
+                return True
+            elif "certificate verify failed" in output:
+                cmd = "sed -i -e 's|^insecure.*|insecure = 1|g' " \
+                      "/etc/rhsm/rhsm.conf"
+                _, _ = ssh.runcmd(cmd)
+        raise FailException(f'Failed to register {host}')
+
+    def unregister(self, guest=False):
+        """
+        Uregister for virt-who host or guest by subscription-manager.
+        :param guest: unregister for guest when define True, virt-who
+            host as default.
+        :return: True of raise fail.
+        """
+        ssh = self.ssh(guest=guest)
+        host = self.host(guest=guest)
+        for i in range(3):
+            _, _ = ssh.runcmd('subscription-manager unregister')
+            ret, output = ssh.runcmd('subscription-manager clean')
+            if ret == 0:
+                logger.info(f'Succeeded to unregister {host}')
+                return True
+            logger.warning(
+                f'Failed to unregister {host}, try again after 180s.')
+            time.sleep(180)
+        raise FailException(f'Failed to unregister {host}')
+
+    def register_command(self, second_org=False, activ_key=False):
+        """
+        Define subscription-manager command for registering based on the
+        configured parameters.
+        :param second_org: extra org defined only for satellite, use
+            default org as default.
+        :param activ_key: activation_key
+        :return: command string
+        """
+        org = self.default_org
+        if self.register_type == 'satellite':
+            if second_org:
+                org = self.secondary_org
+            if activ_key:
+                ack = self.activation_key
+                cmd = f'subscription-manager register ' \
+                      f'--org="{org}" ' \
+                      f'--activationkey="{ack}"'
+                return cmd
+        cmd = f'subscription-manager register ' \
+              f'--username={self.username} ' \
+              f'--password={self.password} ' \
+              f'--org={org}'
+        return cmd
+
+    def rhsm_conf_set(self, guest=False):
+        """
+        Configure the /etc/rhsm/rhsm.conf before registering.
+        :param guest: configure for guest when guest=True, default is
+            virt-who host.
+        :return: None
+        """
+        ssh = self.ssh(guest=guest)
+        _, _ = ssh.runcmd(f'rm -f {self.rhsm_conf};'
+                          f'cp {self.rhsm_conf_backup} {self.rhsm_conf}')
+        if self.register_type == 'rhsm':
+            cmd = "sed -i -e 's|^hostname.*|hostname = {0}|g' " \
+                  "/etc/rhsm/rhsm.conf".format(self.server)
+            _, _ = ssh.runcmd(cmd)
```

Comment:
is it possible without the rhsm.conf backup ...

Reply:
Yeah, after change to register with ```--serverurl``` option, we can skip to configure the rhsm.conf file. We just need to recover all options if need to configure the rhsm.con file in cases, like proxy related cases.